### PR TITLE
upgrade-manager-v2: Load test fixes

### LINKED ISSUE #244: upgrade-manager-v2: maxUnavailable was 1 but the batch size was 2
The spec clearly says the maxUnavailable is 1
`"maxUnavailable":"1"`

But the batch size is 2.
`rotating batch	{"instance IDs": ["i-000d2b376ce24145f", "i-004f192cc63985fcb"],`

```
2021-06-02T02:28:18.088Z	INFO	controllers.RollingUpgrade	admitted new rollingupgrade	{"scalingGroup": "ip-fmea2-ppd-usw2-k8s-instance-manager-nodes", "update strategy": {"type":"randomUpdate","mode":"eager","maxUnavailable":"1","drainTimeout":900}, "name": "instance-manager/rolling-upgrade-nodes-20210602022806"}
2021-06-02T02:28:18.567Z	INFO	controllers.RollingUpgrade	scaling group details	{"scalingGroup": "ip-fmea2-ppd-usw2-k8s-instance-manager-nodes", "launchConfig": "ip-fmea2-ppd-usw2-k8s-instance-manager-nodes-20210602022808", "name": "instance-manager/rolling-upgrade-nodes-20210602022806"}
2021-06-02T02:28:18.567Z	INFO	controllers.RollingUpgrade	checking if rolling upgrade is completed	{"name": "instance-manager/rolling-upgrade-nodes-20210602022806"}
2021-06-02T02:28:18.567Z	INFO	controllers.RollingUpgrade	launch definition differs	{"instance": "i-000d2b376ce24145f", "name": "instance-manager/rolling-upgrade-nodes-20210602022806"}
2021-06-02T02:28:18.567Z	INFO	controllers.RollingUpgrade	selecting batch for rotation	{"name": "instance-manager/rolling-upgrade-nodes-20210602022806"}
2021-06-02T02:28:18.568Z	INFO	controllers.RollingUpgrade	rotating batch	{"instance IDs": ["i-000d2b376ce24145f", "i-004f192cc63985fcb"], "name": "instance-manager/rolling-upgrade-nodes-20210602022806"}
```

### PR DESCRIPTION
**Pull request by shreyas-badiger**
Fixes https://github.com/keikoproj/upgrade-manager/issues/220
Fixes #244 
Fixes https://github.com/keikoproj/upgrade-manager/issues/246
Fixes https://github.com/keikoproj/upgrade-manager/issues/247
Add appropriate logs.

### CODE REVIEW

File: controllers/rollingupgrade_controller.go [hunk 1]
```diff
@@ -65,12 +67,13 @@ type RollingUpgradeAuthenticator struct {
...
 // and the details in the RollingUpgrade.Spec
 func (r *RollingUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Info("***Reconciling***")
 	rollingUpgrade := &v1alpha1.RollingUpgrade{}
 	err := r.Get(ctx, req.NamespacedName, rollingUpgrade)
 	if err != nil {
 		if kerrors.IsNotFound(err) {
-			r.AdmissionMap.Delete(req.NamespacedName)
-			r.Info("deleted object from admission map", "name", req.NamespacedName)
+			r.AdmissionMap.Delete(fmt.Sprintf("%s", req.NamespacedName))
```
Comment (shreyas-badiger) [inline controllers/rollingupgrade_controller.go:75]: without this, the rolling upgrade won't be deleted from the admission map

File: controllers/rollingupgrade_controller.go [hunk 2]
```diff
@@ -120,8 +123,13 @@ func (r *RollingUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return ctrl.Result{RequeueAfter: time.Second * 30}, nil
 	}
 
-	r.Info("admitted new rollingupgrade", "name", rollingUpgrade.NamespacedName(), "scalingGroup", scalingGroupName)
-	r.AdmissionMap.Store(rollingUpgrade.NamespacedName(), scalingGroupName)
+	// store the rolling upgrade in admission map
+	if _, present := r.AdmissionMap.LoadOrStore(rollingUpgrade.NamespacedName(), scalingGroupName); present == false {
+		r.Info("admitted new rolling upgrade", "scalingGroup", scalingGroupName, "update strategy", rollingUpgrade.Spec.Strategy, "name", rollingUpgrade.NamespacedName())
+		r.CacheConfig.FlushCache("autoscaling")
```
Comment (shreyas-badiger) [inline controllers/rollingupgrade_controller.go:129]: This is to make sure the cache flush everytime we admit a new rolling upgrade

File: controllers/rollingupgrade_controller.go [hunk 3]
```diff
@@ -141,17 +149,21 @@ func (r *RollingUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+		r.Info("failed to discover the cloud", "scalingGroup", scalingGroupName, "name", rollingUpgrade.NamespacedName())
 		rollingUpgrade.SetCurrentStatus(v1alpha1.StatusError)
-		// Set prometheus metric cr_status_failed
 		common.SetMetricRollupFailed(rollingUpgrade.Name)
 		return ctrl.Result{}, err
 	}
 
 	// process node rotation
+	scalingGroup := awsprovider.SelectScalingGroup(rollupCtx.RollingUpgrade.ScalingGroupName(), rollupCtx.Cloud.ScalingGroups)
+	r.Info(
```
Comment (shreyas-badiger) [inline controllers/rollingupgrade_controller.go:160]: scaling group details is essential to know if the launch config changed.

File: controllers/upgrade.go [hunk 4]
```diff
@@ -326,12 +324,16 @@ func (r *RollingUpgradeContext) SelectTargets(scalingGroup *autoscaling.Group) [
 
 	var unavailableInt int
 	if batchSize.Type == intstr.String {
-		unavailableInt, _ = intstr.GetValueFromIntOrPercent(&batchSize, totalNodes, true)
+		if strings.Contains(batchSize.StrVal, "%") {
```
Comment (shreyas-badiger) [inline controllers/upgrade.go:327]: To make sure we call `GetValueFromIntOrPercent` only when the value is a percentage.

File: controllers/upgrade.go [hunk 5]
```diff
@@ -91,13 +92,22 @@ func (r *RollingUpgradeContext) RotateNodes() error {
 	var (
 		scalingGroup = awsprovider.SelectScalingGroup(r.RollingUpgrade.ScalingGroupName(), r.Cloud.ScalingGroups)
 	)
+	if reflect.DeepEqual(scalingGroup, &autoscaling.Group{}) {
+		r.Info("scaling group not found", "scalingGroup", r.RollingUpgrade.ScalingGroupName(), "name", r.RollingUpgrade.NamespacedName())
+		return errors.Errorf("scaling group not found")
```
Comment (eytan-avisror) [inline controllers/upgrade.go:97]: this will double log, can you just return an error with all the information and remove the info?

**Review by eytan-avisror — APPROVED** [review]
Minor comment, looks good otherwise

File: controllers/upgrade.go (hunk 5 above)
Comment (shreyas-badiger) [reply controllers/upgrade.go:97]: Yup. That makes sense.

**Pull request merged**